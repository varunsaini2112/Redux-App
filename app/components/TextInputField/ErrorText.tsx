import { Text, View } from "react-native";

import styles from "./TextInputField.styles";

const ErrorText = ({ errorText = "" }: { errorText?: string }) => {
	return (
		<View style={styles.errorContainer}>
			<Text style={styles.errorText}>{errorText}</Text>
		</View>
	);
};

export default ErrorText;
