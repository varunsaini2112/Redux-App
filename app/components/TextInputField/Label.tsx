import { Text, View } from "react-native";

import styles from "./TextInputField.styles";

const Label = ({ label }: { label: string }) => {
	return (
		<View style={styles.labelContainer}>
			<Text style={styles.labelText}>{label}</Text>
		</View>
	);
};

export default Label;
