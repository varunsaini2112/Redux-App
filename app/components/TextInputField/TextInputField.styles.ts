import { StyleSheet } from "react-native";
import { Padding, colors, fontSizes } from "../../config/appStyles";

const styles = StyleSheet.create({
	mainContainer: {
		marginBottom: 16
	},
	labelContainer: {
		marginHorizontal: Padding.s
	},
	labelText: {
		fontSize: fontSizes.m,
		color: colors.text_heading
	},
	textInputField: {
		borderWidth: 0.4,
		padding: 16,
		borderRadius: 16
	},
	errorContainer: {
		alignItems: "flex-end"
	},
	errorText: {
		fontSize: fontSizes.s,
		color: "red"
	}
});

export default styles;
