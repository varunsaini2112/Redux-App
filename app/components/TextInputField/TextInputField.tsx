import { Text, TextInput, TextInputProps, View } from "react-native";
import {
	FieldPath,
	FieldValues,
	PathValue,
	UseControllerProps,
	useController
} from "react-hook-form";
import { Padding, colors, fontSizes } from "../../config/appStyles";

type TextInputFieldProps<
	TFieldValues extends FieldValues = FieldValues,
	TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
> = UseControllerProps<TFieldValues, TName> &
	Omit<TextInputProps, "onChange" | "onBlur" | "value" | "ref"> & {
		label?: string;
	};

const TextInputField = <
	TFieldValues extends FieldValues = FieldValues,
	TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
>(
	props: TextInputFieldProps<TFieldValues, TName>
) => {
	const { label, name, control, rules, ...textInputProps } = props;
	const {
		field,
		formState: { errors }
	} = useController({
		control,
		name,
		defaultValue: "" as PathValue<TFieldValues, TName>,
		rules
	});

	const { value, onChange, onBlur, ref } = field;

	return (
		<View style={{ borderWidth: 0, marginBottom: 16 }}>
			{label && (
				<View style={{ marginHorizontal: Padding.s }}>
					<Text style={{ fontSize: fontSizes.m, color: colors.text_heading }}>
						{label}
					</Text>
				</View>
			)}
			<TextInput
				ref={ref}
				value={value}
				onBlur={onBlur}
				onChangeText={onChange}
				style={{ borderWidth: 0.4, padding: 16, borderRadius: 16 }}
				{...textInputProps}
			/>
			<View style={{ alignItems: "flex-end" }}>
				<Text style={{ fontSize: fontSizes.s, color: "red" }}>
					{(errors[name]?.message as string) ?? ""}
				</Text>
			</View>
		</View>
	);
};

export default TextInputField;
