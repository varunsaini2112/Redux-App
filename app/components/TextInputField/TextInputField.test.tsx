import { Pressable, Text } from "react-native";
import { useForm } from "react-hook-form";
import { act, fireEvent, render, screen } from "@testing-library/react-native";

import TextInputField from "./TextInputField";

const TestComponent = ({
	onPressSubmit = () => {}
}: {
	onPressSubmit?: () => void;
}) => {
	const { handleSubmit, control } = useForm<{ ExampleField: string }>();

	return (
		<>
			<TextInputField
				label="Example Field"
				name="ExampleField"
				placeholder="exampleFieldPlaceholder"
				control={control}
				rules={{ required: { value: true, message: "Field is required" } }}
			/>
			<Pressable onPress={handleSubmit(onPressSubmit)}>
				<Text>Press me</Text>
			</Pressable>
		</>
	);
};

describe("TextInputField Component", () => {
	it("should match snapshots", () => {
		const snapshot = render(<TestComponent />);

		expect(snapshot).toMatchSnapshot();
	});

	it("should render label correctly", () => {
		render(<TestComponent />);

		expect(screen.getByText("Example Field")).toBeOnTheScreen();
	});

	it("should render required error text on initially form submit", async () => {
		render(<TestComponent />);

		await act(async () => {
			fireEvent.press(screen.getByText("Press me"));
		});

		expect(screen.getByText("Field is required")).toBeOnTheScreen();
	});

	it("should call onPress on submit form with correctly field entries", async () => {
		const onPressMock = jest.fn();
		render(<TestComponent onPressSubmit={onPressMock} />);

		fireEvent.changeText(
			screen.getByPlaceholderText("exampleFieldPlaceholder"),
			"randomEntry"
		);

		await act(async () => {
			fireEvent.press(screen.getByText("Press me"));
		});

		expect(onPressMock).toHaveBeenCalled();
	});
});
