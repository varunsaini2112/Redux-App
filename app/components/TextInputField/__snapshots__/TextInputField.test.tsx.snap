// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputField Component should match snapshots 1`] = `
[
  <View
    style={
      {
        "marginBottom": 16,
      }
    }
  >
    <View
      style={
        {
          "marginHorizontal": 8,
        }
      }
    >
      <Text
        style={
          {
            "color": "#132631",
            "fontSize": 14,
          }
        }
      >
        Example Field
      </Text>
    </View>
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="exampleFieldPlaceholder"
      style={
        {
          "borderRadius": 16,
          "borderWidth": 0.4,
          "padding": 16,
        }
      }
      value=""
    />
    <View
      style={
        {
          "alignItems": "flex-end",
        }
      }
    >
      <Text
        style={
          {
            "color": "red",
            "fontSize": 12,
          }
        }
      />
    </View>
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text>
      Press me
    </Text>
  </View>,
]
`;
